### ISSUE TITLE
Custom PlaceholderReplacer Implementations

### ISSUE THREAD

**Issue by mirjalali** [flyway/flyway#1082]
Users of Flyway should be able to use their own implementations of `PlaceholderReplacer`.
This can easily be done by extending the class `Flyway` and overriding the creating method.
No such PlaceholderReplacer-creating method exists in Flyway-3.0.
Fortunately, in source codes of Flyway-4.0, there is a `createPlaceholderReplacer()` method, but it is private!

**Comment by axelfontaine** [comment]
What exactly is missing for you in the current standard placeholder replacer?

Cheers
Axel

**Comment by mirjalali** [comment]
Well in fact it's too static! We are looking for a much more dynamic one. I have the idea of implementing the `PlaceholderReplacer` such that java-callbacks and even some simple expressions are also supported. Imagine the following sql migration (`id` is a simple integer primary key, and `hash_str` is a long-enough randomly-generated string):
```sql
${ id := getMaxId('BOOKS')+1 };
insert into BOOKS(id, hash_str, name)
  values(${id++}, ${genHash()}, 'book1'), (${id++}, ${genHash()}, 'book2');
```
Currently, such migrations can only be implemented using java codes, but I think such a cute feature can be provided using some expression-parsing, dependency injection, and GoF design patterns like chain of responsibility.

**Comment by novotnyr** [comment]
There is an additional use-case for protected `createPlaceholderReplacer()`: supporting regexp-ed prefixes and suffixes.

If you support '@' or '@@' as a prefix, without suffix, you can create the following constructs:

INSERT INTO table VALUES (id, @@name).

This is very similar to MySQLish variable naming, and you can get autocomplete for variables in IntelliJ IDEA for free.

However, this requires empty suffix or allowing regexps in prefixes and suffixes.

**Comment by spt-developer** [comment]
Excuse me for jump in.

I want to create custom `PlaceholderReplacer` too.

I created `SqlConverter` (extends `PlaceholderReplacer`) to execute MySQL-SQL in H2 database.
[SqlConverter.java](https://github.com/spt-oss/flyway-plus/blob/master/src/main/java/org/flywaydb/plus/internal/util/SqlConverter.java)
[MySqlToH2Converter.java](https://github.com/spt-oss/flyway-plus/blob/master/src/main/java/org/flywaydb/plus/internal/util/MySqlToH2Converter.java)

And put it into Flyway by force :)
[CustomFlyway.java](https://github.com/spt-oss/flyway-plus/blob/master/src/main/java/org/flywaydb/core/CustomFlyway.java)

I know it is different to intention of replacer, but there is a demand.

**Comment by lbreuss** [comment]
+1. Good suggestions form @mirjalali.

We need to process sqlplus scripts like this:
`INSERT INTO TD_&tenantId . MODEL (ID,VERSION,INSTALLDATE) VALUES (TD_&tenantId .MODEL_SEQ.NEXTVAL,'flyway uses &tenantId',sysdate);`

In addition to @novotnyr, v4.0 does no longer allow empty placeholderSuffix -- which was sort of a workaround in v3.1.

Suggestion: With a custom PlaceholderReplacer understanding regex as prefix and suffix, we could directly use these sqlplus scripts. Additionally, the placeholder match itself in `checkForUnmatchedPlaceholderExpression()` should not match _any_ character as in "(.+?)", but rather "valid" characters of placeholder identifiers, e.g. "([\\w]+?)", or a configurable regex pattern **`placeholderPattern`**. 

All this could be activated by another setting e.g. **`placeholderReplacementClass=org.flywaydb.core.util.PlaceholderReplacerRegex`**, or even a comma separated list of replacers.

**Comment by axelfontaine** [comment]
@lbreuss The case you describe will be implemented as part of #1883

**Comment by MikielAgutu** [comment]
I think many of the problems raised here have been addressed.
